ENH: Added a new Formatter to mpl.ticker and some example recipes

madphysicist opened the pull request:
- BinaryIntFormatter: Formats ticks as binary integers. Intended for use with `MaxNLocator(integer=True)`.
- HexIntFormatter: Formats ticks as hexadecimal integers. Intended for use with `MaxNLocator(integer=True)`.
- LinearScaleFormatter: Wraps any other formatter and transforms the input to it on a linear scale. The wrapped formatter can be any callable that accepts `x` and `pos` as arguments, but actual `Formatter` instances will get special treatment.

Also, changed the annotation in PercentFormatter tests slightly (non-functional change).

NelleV commented:
What would be the use case for the binary formatter and the hex formatter?

madphysicist commented:
I developed the Binary and Hex formatters specifically for plotting the output of a Digital-to-Analog Converters versus the input. The x-axis needed to indicate the input bits in a way that was easier to understand than decimal numbers.

LinearScaleFormatter is a way of transforming the values shown on the axis without modifying the underlying data. It can probably be generalized to an arbitrary tick value-to-display value mapping function, but I found it useful to just specify a linear transformation with two points on the input range and two points on the output range.

A sample use-case would be if you had data acquired with say imperial units like quarts, but wanted to display the axis in liters. If you did not want to provide a transformed dataset, you could wrap your existing formatter with something like `LinearScaleFormatter(1.0, 0.946353, ax.xaxis.get_major_formatter())` (1qt = 0.946353L). I expect this is much more useful than the other two formatters. I will work on a piece of example code to add to the gallery.

QuLogic commented:
Both the binary and hex formatters could be implemented with (or replaced by?) the `StrMethodFormatter`.

The example of changing units is something we should get more unit support for, though I feel like much of the underlying infrastructure is there already, but I'm not sure how to use it. Maybe one of the unit examples would be useful.

madphysicist commented:
> Both the binary and hex formatters could be implemented with (or replaced by?) the StrMethodFormatter.

I thought so too at first, but it turns out you have to explicitly cast to an int to use `b` and `x` formats. The input to `Formatter.__call__` is pretty much always a `float`, even with `MaxNLocator(integer=True)`.

QuLogic commented:
Really strange that it's not defined because you can do `123.0.hex()` and get `'0x1.ec00000000000p+6'` back.

madphysicist commented:
Yeah. Python 3.5.2:
```
>>> '{:b}'.format(123.0)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: Unknown format code 'b' for object of type 'float'
>>> '{:x}'.format(123.0)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: Unknown format code 'x' for object of type 'float'
```

madphysicist commented:
What I can do is redefine `BinaryIntFormatter` and `HexIntFormatter` as a single class and add a paramter `type`, which would be `'hex'`, `'bin'` or `'oct'`.

madphysicist commented:
@QuLogic I have converted `BinaryIntFormatter` and `HexIntFormatter` into a single extension of `StrMethodFormatter` that mainly just converts the input to an integer before calling the super implementation. Thanks for inspiring that idea.

@NelleV Hopefully the new form is more acceptable. While I agree that the original implementation was a bit domain specific, a formatter that handles the conversion to integer properly is much more generally useful.

I will get the bugs out shortly.

anntzer commented:
For the sake of avoiding adding more very thin classes, I'd suggest simply adding a doc entry suggesting the use of `FuncFormatter(lambda x, i: format(int(x), fmt))`.  This is honestly not that much longer to write than `IntegerFormatter(fmt)`, and *much* clearer for the reader.  Likewise the proposed `unsigned` kwarg can trivially be implemented by using the `-` flag in the format string.  So I'm fairly strongly -1 on adding such a class.

`LinearScaleFormatter` is more promising, but should probably be implemented as a general Formatter transformer that simply takes an arbitrary input->output transform callable.  Also, following @QuLogic's suggestion I'd like first to know whether the unit machinery is already enough (or close to enough, in which case it should be improved accordingly) to handle this case.

In any case, please make all attributes of these new classes private, unless you have a strong reason to keep them public.  Also, please check how `LinearScaleFormatter` interacts with `format_cursor_data`.

madphysicist commented:
@anntzer. While I agree that `IntegerFormatter` is a pretty thin wrapper and can generally be omitted, I am not sure that the `unsigned` keyword can be replaced with a `-` format (which is the default anyway). In fact, it is part of the reason that I created the class in the first place. `unsigned = True` does not just remove the minus sign from the number: it causes negative numbers to be treated as n-bit unsigned numbers. Either way, I am fine with making this a recipe in the examples if that is the general concensus.

I will rewrite `LinearFormatter` as `MappingFormatter`. The linear transformation can then go into a recipe in the examples section.

Finally, when you say "making the attributes private", do you mean prepending an underscore or something else?

story645 commented:
Not sure how the unit infrastructure would help since `unit` is a bad name for it. The infrastructure is really about registering and working with datatypes and it sounds like this data is in floats/ints.

anntzer commented:
Oh, I completely misunderstood what "unsigned" does, sorry about that.  It's arguably a bit more complicated but this can still be implemented as
```
FuncFormatter(lambda x, i:
    format(int.from_bytes(int(x).to_bytes(int(x).bit_length() // 8 + 1, "big", signed=True),
                          "big"),
           fmt)
```
which is not necessarily more legible, but now the argument is instead that this feature may be a bit too specialized?  (Again I'm all for adding it in the docs.)

And yes I meant that attributes must be prepended with an underscore.  Otherwise it basically becomes impossible later to modify the class while maintaining backcompatibility, as users may be relying on whatever public attributes are present...

madphysicist commented:
OK. I will make all the object attributes private. I would like to keep the `transform` method public in all the cases though.

tacaswell commented on lib/matplotlib/ticker.py:
```diff
@@ -195,8 +204,9 @@
-           'Locator', 'IndexLocator', 'FixedLocator', 'NullLocator',
-           'LinearLocator', 'LogLocator', 'AutoLocator',
+           'LinearScaleFormatter', 'BinaryIntFormatter',
+           'HexIntFormatter', 'Locator', 'IndexLocator', 'FixedLocator',
```
There are non-existent classes in here.

tacaswell commented on lib/matplotlib/ticker.py:
```diff
@@ -154,6 +154,15 @@
 :class:`PercentFormatter`
     Format labels as a percentage
 
+:class:`BinaryIntFormatter`
```
These docs have gotten out of sync with the PR.

madphysicist replied:
That last push was mostly an accident. I am overhauling the code based on @anntzer's advice. Thinking about turning some of it into recipes in the docs.

tacaswell commented on lib/matplotlib/ticker.py:
```diff
@@ -1292,6 +1302,128 @@ def convert_to_pct(self, x):
+    This formatter can use any other formatter to actually render the
+    ticks.
+
+    inRef: number or 2-element iterable
```
nitpick, there should be a space before the `:` to format correctly with numpydoc

tacaswell replied:
The last input to this is already a formatter instance, maybe take in a `Norm` instance as the input here?

If not, I have a slight preference to 4 inputs instead of 2 of mixed type.

madphysicist replied:
`Norm` instance?

tacaswell commented on lib/matplotlib/ticker.py:
```diff
@@ -1292,6 +1302,128 @@ def convert_to_pct(self, x):
+
+        # All these values are retained for debugging/extension.
+        # Only the minima are used explicitly.
+        self.iMin, self.iMax = unpack(inRef, 'in')
```
The style tends to be camel case for classes, `_` for everything else.

madphysicist commented:
@tacaswell I believe that my latest push addresses your immediate concerns.

madphysicist commented:
@anntzer , in reference to your [request](https://github.com/matplotlib/matplotlib/pull/7482#issuecomment-261699810)

> Also, please check how LinearScaleFormatter interacts with format_cursor_data.

`format_cursor_data` is defined in `Artist`, and does not get overriden by `Axes` or `Axis` as far as I can tell. There is no direct interaction with a formatter at all.

If you meant the default implementation of [`_AxesBase.format_coord`](https://github.com/matplotlib/matplotlib/blob/57c845b158df87a23635b39447647b8e404cbdbf/lib/matplotlib/axes/_base.py#L3373), it does in fact [use the major formatter's `format_data_short` method](https://github.com/matplotlib/matplotlib/blob/57c845b158df87a23635b39447647b8e404cbdbf/lib/matplotlib/axes/_base.py#L3356). I will therefore go ahead and made proper references to the remaining methods of the underlying `Formatter`.

madphysicist commented:
I took all of the advice I got on this thread and made a completely new class called `TransformFormatter` which allows generic mapping of tick values before plugging them into a different Formatter or formatting function. I then added the two Formatters I had before to an example in the documentation since they are now trivial to implement. The new tests verify, among other things, that `format_data_short` works correctly.

The only major comment that I am no longer following here is that the two attributes of this new class are both public. I think that in this context it makes sense to do it this way, in the same way that `FormatStrFormatter` exposes the `fmt` attribute.

Another possible problem is that this class is sensitive to changes in the underlying API. This is probably fine, but I want to add a note to add the appropriate redirects in `TransformFormatter` if the `Formatter` or `TickHelper` methods ever change. Where would I add such a note to future developers?

Finally, I have a purely technical question. At the moment, the `axis` attribute is a reference shared by the base instance and the underlying formatter. However, this means that the view and data bounds are not transformed for the underlying formatter. Should I make the `formatter` attribute always have a `_DummyAxis` assigned to it with the transformed bounds? Or is the way I currently did it going to work OK?

madphysicist commented:
Could someone please explain why the appveyor test is failing while the same configuration in Travis is passing? Is it something I am missing that can be easily fixed or should I change the test to avoid using unicode strings?

tacaswell commented:
@madphysicist Because windows ;)

tacaswell commented on lib/matplotlib/ticker.py:
```diff
@@ -1292,6 +1295,117 @@ def convert_to_pct(self, x):
+        # only evaluate this if necessary
+        default = getattr(super(TransformFormatter, self), name)
+        retval2 = default(*args, **kwargs)
+        return retval if self._need_redirect else retval2
```
Huh, I am surprised that this does not raise a `NameError` but it makes sense that it does not.

madphysicist replied:
I am not sure why this is surprising. I always call this with the names of existing methods.

tacaswell commented on lib/matplotlib/ticker.py:
```diff
@@ -1292,6 +1295,117 @@ def convert_to_pct(self, x):
+        return retval if self._need_redirect else retval2
+
+    def set_axis(self, ax):
+        self._invoke_redirect('set_axis', True, ax)
```
I think this is the only `both=True`?  Slight preference to special case here and keep `_invoke_redirect` simpler.

NelleV commented:
This is almost ready to be merged. @madphysicist, do you think you can do the small changes @tacaswell asked for?

madphysicist commented:
@NelleV I have been a bit short on time for the past couple of weeks. Will fix soon.

madphysicist commented:
Nominally, I made the change that @tacaswell requested. I factored out the two different behaviors into different functions, although the net result may be quite a bit more complicated than it was before. This is a factor of added functionality, though, so hopefully you approve.

The main change is that I updated the way that the axis is handled in the underlying formatter. I ended up always setting up a dummy axis with the bounds transformed relative to the actual axis. This allows the underlying formatter to interpret numbers correctly if it cares about the bounds. It also allows the bounds to be of a type other than `float` if that is what the transformer outputs/formatter requires.

The changes are significant enough where I think this needs some re-review.

story645 commented on examples/ticks_and_spines/tick_transform_formatter.py:
```diff
@@ -0,0 +1,128 @@
+)
+
+
+class LinearTransform:
```
While I'm in awe of all the documentation, I wonder if in this case it obscures the use case for this? This is fairly complicated, so at the least is there a simpler 10-20 line use case for the intermediate user that you can turn into a secondary example?

madphysicist replied:
The origin behind this file is that the original version of the formatter was doing this linear transformation. I was asked to make the formatter more general and provide the linear transformation as a recipe instead. I have no problem addding a simpler example. Did you have something particular in mind?

story645 replied:
Not really, my only request is that it be as few LOC as possible while still being clear and accessible to a wider audience. Maybe a temperature convertor?

madphysicist replied:
Just to be clear, you are OK with keeping this particular bit of code around (perhaps in a different location), but would like to see something simpler as the basic example of the new formatter?

story645 replied:
Exactly!

story645 commented on lib/matplotlib/tests/test_ticker.py:
```diff
@@ -511,37 +511,147 @@ def test_formatstrformatter():
-                         percentformatter_test_cases)
+    [
+        # Check explicitly set decimals over different intervals and values
+        (100, 0, '%', 120, 100, '120%'),
```
can you add [ids](http://docs.pytest.org/en/latest/fixture.html#parametrizing-fixtures)? It's really hard to distinguish what different things each of these 5 (or other grouped) cases are trying to test. (And if they're not testing distinct things, then they should be cut.

madphysicist replied:
Yes, will fix.

story645 replied:
:)

madphysicist replied:
Does running `python tests.py` not use `pytest` properly for some reason?

QuLogic replied:
No, that has not yet been changed from using `nose`; use `pytest` directly in the meantime.

story645 commented on lib/matplotlib/tests/test_ticker.py:
```diff
@@ -511,37 +511,147 @@ def test_formatstrformatter():
     assert formatter.format_pct(x, display_range) == expected
 
 
+def test_transformformatter():
```
Should this be a class? pretty much everything followed by a # should be its own. It's at the bottom of the page here: http://docs.pytest.org/en/latest/fixture.html?highlight=classes

story645 commented on lib/matplotlib/ticker.py:
```diff
@@ -1361,6 +1378,186 @@ def convert_to_pct(self, x):
+
+    def _itransform(self, arg):
+        """
+        Transforms an iterables element-by-element, returns a list.
```
typo: iterable

madphysicist replied:
Again, remnant of older discarded version. Will fix.

madphysicist commented:
@story645 Thanks for the detailed review. I have fixed everything you mentioned:

  - Added a simplified example to the great big formatter example
  - Added IDs to `test_percentformatter`
  - Classified `test_transformformatter`
  - Fixed the typo
  - Fixed PEP8 errors that were making the tests fail
  - Allowing `ScalarFormatter` to do `set_locs(None)` like all the other formatters without crashing.

I made one additional change to the big formatter example besides adding my code:
```
ax.xaxis.set_major_locator(ticker.AutoLocator())
ax.xaxis.set_minor_locator(ticker.AutoMinorLocator())
```
for `ScalarFormatter` was changed to
```
ax.xaxis.set_major_locator(ticker.MultipleLocator(1.00))
ax.xaxis.set_minor_locator(ticker.MultipleLocator(0.25))
```
which is what all the other examples use. It does not appear to have altered the output in any way, but should probably be reviewed anyway.

How did you make the comment "Also not totally convinced PercentFormatter needs to exist."? I seem to be unable to reply to it inline. `PercentFormatter` comes from a previous PR. The main justification for it is that it appears to be a very common use case. I wrote it in response to the following two questions: http://stackoverflow.com/q/26294360/2988730, http://stackoverflow.com/q/31357611/2988730 as well as my own needs. While I agree that it probably needs a bit more work, I think that the fundamental concept is sound and makes sense in matplotlib.

madphysicist commented:
I believe that I have fixed all the remaining bugs, including the Linux/Windows minus-sign mixup. @tacaswell If you are happy with the way I handled the original review comments you made, could you close that out? I think this PR is ready for merge barring additional requests.

story645 commented on doc/users/whats_new/new_formatters.rst:
```diff
@@ -0,0 +1,30 @@
+
+A more generic version of :class:`matplotlib.ticker.FuncFormatter` that
+allows the tick values to be transformed before being passed to an
+underlying formatter. The transformation can yeild results of arbitrary
```
spelling: yield

story645 commented on examples/ticks_and_spines/tick-formatters.py:
```diff
@@ -64,29 +60,31 @@ def major_formatter(x, pos):
+ax = plt.subplot(n, 1, 7)
+setup(ax)
+ax.xaxis.set_major_formatter(ticker.TransformFormatter(lambda x: 7 - 2 * x))
+ax.text(0.0, 0.1, "TransformFormatter(lambda x: 7 - 2 * x)",
```
😄 though I don't really grok the difference between TransformFormatter vs. FuncFormatter...is it a data space vs. axis space thing?

madphysicist replied:
The original idea stems from trying to use `"{x:x}"` as an argument to `StrMethodFormatter`, which does not work. But now you can do
```
TransformFormatter(transform=lambda x: int(x), formatter=StrMethodFormatter())`
```
and everything works out just fine. More details starting with [this comment](https://github.com/matplotlib/matplotlib/pull/7482#issuecomment-261440258).

madphysicist replied:
Also, if you want to just switch physical units in your data, you can set the transform, but still use the formatter you like without having to rewrite the whole thing. `ScalarFormatter.__call__` depends on the bounds of the axis. If you switch units, you'd want the bounds it is working with to be in the units of the transformed data, not the actual. That is basically what `TransformFormatter` does.

madphysicist replied:
I think most of this is covered in the docs I wrote, but I can try to contrive an example if you still think that's necessary. It certainly never hurts to have more documentation.

story645 replied:
Actually too much documentation can get overwhelming; the goal is the right sorta documentation. I'm half tempted to let it go though as it's probably not worth holding the PR over the mini-example.

madphysicist replied:
That works for me, but I would really like #7993 to go first (or get rejected) if that's OK with you.

madphysicist replied:
Here is an example of where the linear transformation recipe I made would actually be helpful: http://stackoverflow.com/q/41984104/2988730

story645 commented on lib/matplotlib/tests/test_ticker.py:
```diff
@@ -511,37 +511,223 @@ def test_formatstrformatter():
-)
-
-
 @pytest.mark.parametrize('xmax, decimals, symbol, x, display_range, expected',
```
this is purely style, but I prefer to leave the test_cases and id out of the decorator (the way you had before) and just pass in the variables (just as you had) instead of creating the super long decorator)
here's a teeny example: https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/tests/test_category.py#L17

but I am thrilled about the ids

madphysicist replied:
I prefer that way too. I was mostly just trying to stay consistent with the other tests in the module, but it's probably better to clean them up as well. I will do this in a separate PR.

madphysicist replied:
Done in #7993

story645 commented on lib/matplotlib/ticker.py:
```diff
@@ -242,6 +245,9 @@ def set_view_interval(self, vmin, vmax):
     def get_minpos(self):
         return self._minpos
 
+    def set_minpos(self, minpos=0):
```
since this is new, should it be a [property](https://www.programiz.com/python-programming/property) and should get_minpos be deprecated....? Also, I acknowledge that's very likely out of scope.

madphysicist replied:
This is a module private class that attempts to mimic a couple of the behaviors of `Axis` that are used by formatters and locators. I think that ` get_minpos` is part of that API and shouldn't be touched. I also think that turning `minpos` into a property would cause issues for the same reason. `set_minpos` is a thing I introduced to avoid reallocating a new `_DummyAxis` every time `_minpos` changes because of a transformer change.

story645 replied:
*shrugs* then I'll just write it off as out of scope for this PR and a debatable major overhaul for like a version 3.0

madphysicist commented:
The Travis tests are failing because of something outside this PR. I think I nailed down the last Appveyor error.

madphysicist commented:
Rebased temporarily onto #7965. This is just to make my life easier when #7965 gets merged. Please ignore until then.

madphysicist commented:
As far as I can tell, I have addressed all the issues with this PR. Still waiting on approval from @tacaswell  and at least one other reviewer though.

madphysicist commented:
Bump.

story645 commented on examples/ticks_and_spines/tick_transform_formatter.py:
```diff
@@ -0,0 +1,128 @@
+          - both ends are missing: set ranges to 1.0
+          - one end is missing: set it to the other end
+        """
+        self._in_offset = 0.0 if in_start is None else in_start
```
why aren't you setting in_start=0  and out_start=0 in the __init__(...)?

madphysicist replied:
I want to be able to test if the parameter was set or not. `None` allows me to implement all the combinations listed in the docstring. Setting to a number would not allow that.

madphysicist replied:
Nevermind. I do see your point. Will fix.

story645 commented on examples/ticks_and_spines/tick_transform_formatter.py:
```diff
@@ -0,0 +1,128 @@
+                self._in_scale = out_end - self._in_offset
+        else:
+            self._in_scale = in_end - self._in_offset
+
```
So is this 
```python
if in_end is not None:
    self._in_scale = in_end - self._in_offset
else if out_end is not None:
    self._in_scale = out_end - self._in_offset
else:
    self._in_scale = 1.0
```

story645 commented on examples/ticks_and_spines/tick_transform_formatter.py:
```diff
@@ -0,0 +1,128 @@
+        else:
+            self._in_scale = in_end - self._in_offset
+
+        if out_end is None:
```
can this also be rewritten to remove one layer of nesting?
```python

if out_end is not None:
    self._out_scale = out_end - self._out_offset
else if in_end is not None:
    self._out_scale = in_end - self._out_offset
else:
    self._out_scale = 1.0
```

madphysicist replied:
Took me a minute to realize how you just flipped the `if` so that the `else` comes first. Slow day. I did what you propose except that I kept the `elif` checking for `in_end is None`.

I was thinking that I can actually compute a single scale and offset parameter to minimize the math done in `__call__`. What it your thought on that?

story645 replied:
I ❤️ contrapositives
I like `elif in_end is not` 'cause it keeps the formulas together and so parallel structure, but :shrugs: not gonna hold up a review over it.

madphysicist replied:
I'm not really hung up on positives. Now that I've made the change, it does look better that way. Honestly, I am pretty sure I subconsciously avoid contrapositives after having worked in a place where they were discouraged by the coding standard for no really good reason.

madphysicist replied:
I can rewrite `__call__` as `return x * self._scale + self._offset` by adding probably about two extra lines to `__init__`. Personally I think it's worth precomputing everything in `__init__`, but I am OK with keeping the structure as-is if that is what you prefer.

story645 replied:
I have no preference here, so go with whatever you think is more sensible/maintainable

madphysicist replied:
I will go ahead and make the change. I also just realized that when I removed the lines like `self._in_offset = 0.0 if in_start is None else in_start`, I forgot to replace them with `self._in_offset = in_start` and similar :)

madphysicist replied:
Is there some sort of test that I can run on my example, like an image test?

madphysicist replied:
I made the additional change and ran the code this time to make sure it actually works as intended :)

story645 commented on examples/ticks_and_spines/tick_transform_formatter.py:
```diff
@@ -0,0 +1,128 @@
+
+fig = plt.figure()
+ax1 = fig.add_subplot(111)
+ax2 = fig.add_subplot(111, sharex=ax1, sharey=ax1, frameon=False)
```
Is it necessary to have a shared axis to display this new ticker? That seems to be conflating two separate tasks.

madphysicist replied:
Not strictly necessary, but the conflation was deliberate. I wanted to be able to display both the original and the transformed values as well as to use this as a demo for #7528. There will be some modifications once that PR gets accepted (which I think is very likely once I find the time to write some additional tests for it).

story645 commented on lib/matplotlib/tests/test_ticker.py:
```diff
@@ -613,3 +613,175 @@ def test_latex(self, is_latex, usetex, expected):
+        assert fmt.locs == locs
+        assert fmt.formatter.locs == xlocs
+
+        fmt.set_locs(None)
```
this is  a teeny test, but it's a different one `def test_clear_locs` or something

madphysicist replied:
Done

madphysicist commented:
I don't think the travis error has anything to do with me...

madphysicist commented:
'nother bump.

QuLogic commented on examples/ticks_and_spines/tick_transform_formatter.py:
```diff
@@ -0,0 +1,124 @@
+        `out_start` and `in_end` gets mapped to `out_end`. The following
+        shortcuts apply when only some of the inputs are specified:
+
+          - none: no-op
```
'none' should probably be 'no inputs specified'; I thought this was a typo for `None` at first.

QuLogic commented on examples/ticks_and_spines/tick_transform_formatter.py:
```diff
@@ -0,0 +1,124 @@
+
+          - none: no-op
+          - in_start: translation to zero
+          - out_start: translation from zero
```
For both of these, translation of what?

QuLogic commented on examples/ticks_and_spines/tick_transform_formatter.py:
```diff
@@ -0,0 +1,124 @@
+          - in_start: translation to zero
+          - out_start: translation from zero
+          - in_end: scaling to one (divide input by in_end)
+          - out_end: scaling from one (multiply input by in_end)
```
Should be `out_end`?

QuLogic commented on examples/ticks_and_spines/tick_transform_formatter.py:
```diff
@@ -0,0 +1,124 @@
+          - out_start: translation from zero
+          - in_end: scaling to one (divide input by in_end)
+          - out_end: scaling from one (multiply input by in_end)
+          - in_start, out_start: translation
```
Translation of what?

QuLogic commented on examples/ticks_and_spines/tick_transform_formatter.py:
```diff
@@ -0,0 +1,124 @@
+          - in_end: scaling to one (divide input by in_end)
+          - out_end: scaling from one (multiply input by in_end)
+          - in_start, out_start: translation
+          - in_end, out_end: scaling (in_start and out_start zero)
```
I don't know what the note in the parentheses is trying to say.

QuLogic commented on examples/ticks_and_spines/tick_transform_formatter.py:
```diff
@@ -0,0 +1,124 @@
+          - in_start, out_start: translation
+          - in_end, out_end: scaling (in_start and out_start zero)
+          - in_start, out_end: in_end=out_end, out_start=0
+          - in_end, out_start: in_start=0, out_end=in_end
```
i don't know what these two are trying to say.

QuLogic commented on examples/ticks_and_spines/tick_transform_formatter.py:
```diff
@@ -0,0 +1,124 @@
+        Based on the following rules:
+
+          - start missing: set start to zero
+          - both ends are missing: set ranges to 1.0
```
The term "ranges" has never been mentioned before; what ranges?

QuLogic commented on examples/ticks_and_spines/tick_transform_formatter.py:
```diff
@@ -0,0 +1,124 @@
+        """
+        Sets up the transformation such that `in_start` gets mapped to
+        `out_start` and `in_end` gets mapped to `out_end`. The following
+        shortcuts apply when only some of the inputs are specified:
```
After commenting on each of these elements, I think they may just be more confusing than it's worth. Is it really necessary to describe all the "shortcuts"? Would a mathematical equation and a set of defaults be more straightforward?

madphysicist replied:
I think you are right. I wrote this when I was trying to figure out how I would want the different cases to work out mostly as notes for myself. Will fix.

QuLogic commented on examples/ticks_and_spines/tick_transform_formatter.py:
```diff
@@ -0,0 +1,124 @@
+ax1.yaxis.set_major_formatter(
+        TransformFormatter(int, StrMethodFormatter('{x:02X}')))
+
+ax1.set_xlabel('Temperature (\u00B0C)')
```
@anntzer changed these to all use named entities; please change to `\N{DEGREE SIGN}`.

QuLogic commented on lib/matplotlib/tests/test_ticker.py:
```diff
@@ -613,3 +613,176 @@ def test_latex(self, is_latex, usetex, expected):
+    def transform2(self, x):
+        return 2 * x
+
+    @pytest.fixture()
```
You can drop the parentheses.

QuLogic commented on lib/matplotlib/tests/test_ticker.py:
```diff
@@ -613,3 +613,176 @@ def test_latex(self, is_latex, usetex, expected):
+
+    def test_attributes(self, empty_fmt):
+        # Using == is the right way to compare bound methods:
+        # http://stackoverflow.com/q/41900639/2988730
```
https

madphysicist replied:
Why? Both ways work fine. Is there some reason https is preferred?

madphysicist replied:
Just so we're clear, I'm going to make the change. I was just curious.

QuLogic commented on lib/matplotlib/tests/test_ticker.py:
```diff
@@ -613,3 +613,176 @@ def test_latex(self, is_latex, usetex, expected):
+
+    @pytest.fixture()
+    def fmt(self):
+        fmt = self.empty_fmt()
```
This could take `empty_fmt` as an argument to make it clear it uses the fixture as well.

madphysicist replied:
I did not know you could do that to non-test methods. Will definitely chain it that way.

QuLogic commented on lib/matplotlib/tests/test_ticker.py:
```diff
@@ -613,3 +613,176 @@ def test_latex(self, is_latex, usetex, expected):
+
+    @pytest.fixture()
+    def loc_fmt(self):
+        fmt = self.fmt()
```
Same with `fmt` here.

QuLogic commented on lib/matplotlib/ticker.py:
```diff
@@ -1383,6 +1400,186 @@ def symbol(self):
+        transformed values.
+
+        The input may be an instance of
+        :class:`matplotlib.ticker.Formatter` or an other callable with
```
an other -> another

madphysicist commented:
@QuLogic Based on your comments, I have greatly simplified the configuration of the linear transform in the example. I hope you like it. I found myself trying to figure out where I was going with the original version and just gave up :)

I made all the other changes you recommended as well.

tacaswell commented:
@madphysicist Sorry this has completely fallen through the cracks (for 4 years!).  Are you still interested in working on this?

jklymak commented:
This had a *lot* of input, but I'll close for now as orphaned...
